curvefs/src/metaserver/s3compact_wq_impl.cpp

```diff
@@ -0,0 +1,415 @@
+    };
+
+    for (auto i = 0; i < s3chunkinfolist.s3chunks_size(); i++) {
+        auto chunkinfo = s3chunkinfolist.s3chunks(i);
+        addToValidList(chunkinfo.offset(), chunkinfo.offset() + chunkinfo.len(),
+                       chunkinfo.chunkid(), chunkinfo.compaction(),
+                       chunkinfo.offset(), chunkinfo.zero());
+    }
+
+    // merge same chunkid continuous nodes
+    for (auto curr = validList.begin();;) {
+        auto next = std::next(curr);
+        if (next == validList.end()) break;
+        if (curr->end == next->begin) {
+            if (curr->chunkid == next->chunkid) {
+                next->begin = curr->begin;
+                validList.erase(curr);
+            }
+        }
+        curr = next;
+    }
+    return validList;
+}
+
+int S3CompactWorkQueueImpl::ReadFullChunk(
+    const std::list<struct S3CompactWorkQueueImpl::Node> validList,
+    uint64_t inodeId, uint64_t blockSize, std::string* fullChunk,
+    uint64_t* newChunkid, uint64_t* newCompaction) {
+    for (auto curr = validList.begin(); curr != validList.end();) {
+        auto next = std::next(curr);
+        if (curr->zero) {
+            fullChunk->append(curr->end - curr->begin, '\0');
+            curr = next;
+            continue;
+        }
+        for (uint64_t index = (curr->begin - curr->chunkoff) / blockSize;
+             index * blockSize + curr->chunkoff < curr->end; index++) {
+            // read the block obj
+            std::string objName =
+                GenObjName(curr->chunkid, curr->compaction, index, inodeId);
+            std::string buf;
+            const Aws::String aws_key(objName.c_str(), objName.size());
+            int ret = s3Adapter_->GetObject(aws_key, &buf);
+            if (ret != 0) {
+                return ret;
+            } else {
+                uint64_t blockBegin = index * blockSize + curr->chunkoff;
+                uint64_t blockEnd = (index + 1) * blockSize + curr->chunkoff;
+                if (curr->begin >= blockBegin && curr->end <= blockEnd) {
+                    (*fullChunk) += buf.substr(curr->begin - blockBegin,
+                                               curr->end - curr->begin);
+                } else if (curr->begin >= blockBegin && curr->end > blockEnd) {
+                    (*fullChunk) += buf.substr(curr->begin - blockBegin,
+                                               blockEnd - curr->begin);
+                } else if (curr->begin < blockBegin && curr->end > blockEnd) {
+                    (*fullChunk) += buf.substr(0, blockSize);
+                } else if (curr->begin < blockBegin && curr->end <= blockEnd) {
+                    (*fullChunk) += buf.substr(0, curr->end - blockBegin);
+                }
+            }
+        }
+
+        if (curr->chunkid >= *newChunkid) {
+            (*newChunkid) = curr->chunkid;
+            (*newCompaction) = curr->compaction;
+        }
+        if (next != validList.end() && curr->end < next->begin) {
```

Comment:
I think that if node is zero or incontinuity，no need  to merge

Reply:
we only focus on chunk level data, not chunkinfo level

Reply:
I think that add a switch